### PR TITLE
Update "save stretch to cube" ability.

### PR DESCRIPTION
**Pull request by krlberry**
##  Description
Refactors previous work to add the ability to save a stretch to a cube from qview and adds the ability to save/restore/delete an RGB stretch.

## Related Issue
This was an LROC request with no associated issue.

## Motivation and Context
Adds the ability to save an RGB stretch to a cube. 

## How Has This Been Tested?
Tested by-hand (GUI program) by me. Additional gtests added for the new classes.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Documentation change (update to the documentation; no code change)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] I have read and agree to abide by the [Code of Conduct](https://usgs-astrogeology.github.io/code/)
- [x] I have read the [**CONTRIBUTING**](https://github.com/USGS-Astrogeology/ISIS3/blob/dev/CONTRIBUTING.md) document.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.
- [ ] I have added myself to the [AUTHORS.rst](https://github.com/USGS-Astrogeology/ISIS3/blob/dev/AUTHORS.rst) document.

## Licensing
This project is mostly composed of free and unencumbered software released into the public domain, and we are unlikely to accept contributions that are not also released into the public domain. Somewhere near the top of each file should have these words:

> This work is free and unencumbered software released into the public domain. In jurisdictions that recognize copyright laws, the author or authors of this software dedicate any and all copyright interest in the software to the public domain.

- [x] I dedicate any and all copyright interest in this software to the public domain. I make this dedication for the benefit of the public at large and to the detriment of my heirs and successors. I intend this dedication to be an overt act of relinquishment in perpetuity of all present and future rights to this software under copyright law.

### CODE REVIEW

File: isis/tests/StretchBlobTest.cpp [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+  CubeStretch cubeStretch("CetStretch", "defaultType", 1);
+  StretchBlob stretchStretchBlob(cubeStretch);
+
+  EXPECT_STREQ(stretchBlob.Name().toLatin1().data(), "CubeStretch");
+  EXPECT_STREQ(stretchBlob.Type().toLatin1().data(), "Stretch"); 
+
+  EXPECT_STREQ(nameStretchBlob.Name().toLatin1().data(), "name");
+  EXPECT_STREQ(nameStretchBlob.Type().toLatin1().data(), "Stretch"); 
+
+  EXPECT_STREQ(stretchStretchBlob.Name().toLatin1().data(), "CubeStretch");
```
Comment (ladoramkershner) [inline isis/tests/StretchBlobTest.cpp:36]: Based on line 24 setting the name of cubeStretch to "CetStretch", wouldn't this fail?

File: isis/src/base/objs/Stretch/Stretch.h [hunk 2]
```diff
@@ -90,13 +89,10 @@ namespace Isis {
       double p_minimum; //!<By default this value is set to p_lrs
       double p_maximum; //!<By default this value is set to p_hrs
 
-      QString p_type; //! Type of stretch. This is only currently used in the AdvancedStretchTool.
```
Comment (ladoramkershner) [inline isis/src/base/objs/Stretch/Stretch.h:94]: It seems like in this file you are removing the blob reading/writing capability and you add that to the Blob object, which totally makes sense. I am just curious why you are removing the stretch type property from here?

Comment (krlberry) [reply isis/src/base/objs/Stretch/Stretch.h:94]: :+1: The stretchType (Binary, Sawtooth, Linear, or Manual) property only really makes sense in the context of the Advanced Stretch Tool, and I'd originally added it to the Stretch class the last time I worked on this "Save stretch" project. The Stretch base class is used to represent a stretch pair or series of stretch pairs, and is used in various places in ISIS besides the context of saving/loading stretches from cubes. I added the CubeStretch class in this PR to specifically deal with the information needed in a saved stretch for a cube, which includes the stretch type and a band number.

File: isis/tests/StretchBlobTest.cpp (hunk 1 above)
Comment (krlberry) [reply isis/tests/StretchBlobTest.cpp:36]: Thanks for catching this! I wonder if this was a strange typo I managed to mash in-between testing and pushing the change.

File: isis/src/base/objs/Blob/Blob.cpp [hunk 3]
```diff
@@ -175,8 +175,12 @@ namespace Isis {
    *  blob is stored and the pointer is removed from the blob pvl.
    *  
    *  @param pvl The Pvl to be searched
+   *  @param keywords A list of keyword, value pairs to match inside the blob's
+   *  PVL object. Only if all the keyword match is the blob processed. This is used
+   *  when there are multiple blobs with the same name, but different keywords that 
+   *  define the exact blob (see Stretch with a band number)
```
Comment (jessemapel) [inline isis/src/base/objs/Blob/Blob.cpp:181]: If these are keyword value pairs from PVL, then why not just use a vector of PvlKeywords instead of a map/vector of pairs?

File: isis/src/base/objs/Blob/Blob.cpp [hunk 4]
```diff
@@ -187,9 +191,27 @@ namespace Isis {
           QString curName = obj["Name"];
           curName = curName.toUpper();
           if (blobName == curName) {
-            p_blobPvl = obj;
-            found = true;
-            break;
+
+            if (keywords.size() >= 1) {
```
Comment (jessemapel) [inline isis/src/base/objs/Blob/Blob.cpp]: use `if (!keywords.empty())`

File: isis/src/base/objs/Blob/Blob.cpp [hunk 5]
```diff
@@ -187,9 +191,27 @@ namespace Isis {
...
+                  found = false;
+                }
+                ++i;
+              }
+              if (found) {
+                p_blobPvl = obj;
+                found = true;
+                break;
+              }
+            }
```
Comment (jessemapel) [inline isis/src/base/objs/Blob/Blob.cpp:215]: If at matching Pvl label is not found, what happens?

File: isis/src/base/objs/Cube/Cube.h [hunk 6]
```diff
@@ -257,9 +259,10 @@ namespace Isis {
       void open(const QString &cfile, QString access = "r");
       void reopen(QString access = "r");
 
-      void read(Blob &blob) const;
+      void read(Blob &blob, 
+                const QMap<QString,QString> keywords = QMap<QString,QString>()) const;
```
Comment (jessemapel) [inline isis/src/base/objs/Cube/Cube.h]: I'm not a huge fan of how this stuff trickles into the Cube API from the Blob API. This is mostly a problem with the existing Blob API and not your changes.

File: isis/src/base/objs/CubeStretch/CubeStretch.cpp [hunk 7]
```diff
@@ -0,0 +1,162 @@
...
+ *   software and related material nor shall the fact of distribution
+ *   constitute any such warranty, and no responsibility is assumed by the
+ *   USGS in connection therewith.
+ *
+ *   For additional information, launch
+ *   $ISISROOT/doc//documents/Disclaimers/Disclaimers.html
+ *   in a browser or see the Privacy &amp; Disclaimers page on the Isis website,
+ *   http://isis.astrogeology.usgs.gov, and the USGS privacy and disclaimers on
+ *   http://www.usgs.gov/privacy.html.
+ */
```
Comment (jessemapel) [inline isis/src/base/objs/CubeStretch/CubeStretch.cpp]: This is the old boilerplate see https://github.com/USGS-Astrogeology/ISIS3/blob/dev/isis/templates/default.cpp

File: isis/src/base/objs/CubeStretch/CubeStretch.h [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+ *   software and related material nor shall the fact of distribution
+ *   constitute any such warranty, and no responsibility is assumed by the
+ *   USGS in connection therewith.
+ *
+ *   For additional information, launch
+ *   $ISISROOT/doc//documents/Disclaimers/Disclaimers.html
+ *   in a browser or see the Privacy &amp; Disclaimers page on the Isis website,
+ *   http://isis.astrogeology.usgs.gov, and the USGS privacy and disclaimers on
+ *   http://www.usgs.gov/privacy.html.
+ */
```
Comment (jessemapel) [inline isis/src/base/objs/CubeStretch/CubeStretch.h]: This is also the old boilerplate.

File: isis/src/base/objs/CubeStretch/CubeStretch.h [hunk 9]
```diff
@@ -0,0 +1,72 @@
...
+
+#include "Stretch.h"
+#include "StretchBlob.h"
```
Comment (jessemapel) [inline isis/src/base/objs/CubeStretch/CubeStretch.h]: This isn't used anywhere in the API, it should be included in the source file only.

File: isis/src/base/objs/CubeStretch/CubeStretch.cpp [hunk 10]
```diff
@@ -0,0 +1,162 @@
...
+  /**
+   * Constructs a CubeStretch object with default mapping of special pixel values to
+   * themselves and a provided name, and a provided stretch type
+   *  
+   * @param name Name to use for Stretch 
+   * @param type Type of stretch 
+   */
+  CubeStretch::CubeStretch(QString name, QString stretchType, int bandNumber) : m_name(name), 
+    m_type(stretchType), m_bandNumber(bandNumber) {
+  }
```
Comment (jessemapel) [inline isis/src/base/objs/CubeStretch/CubeStretch.cpp:21]: All three of these constructors can be handled by a singe constructor with default arguments.

```
CubeStretch(QString name ="DefaultStretch" , QString stretchType = "Default", int bandNumber = 1)
```

The last constructor already partially does this.

File: isis/src/base/objs/CubeStretch/CubeStretch.h [hunk 11]
```diff
@@ -0,0 +1,72 @@
...
+  class CubeStretch : public Stretch { 
+    public: 
+      CubeStretch();
+      CubeStretch(QString name);
+      CubeStretch(QString name, QString stretchType, int bandNumber=1);
+      ~CubeStretch();
+
+      CubeStretch(Stretch const& stretch);
+      CubeStretch(Stretch const& stretch, QString type);
+      CubeStretch(Stretch const& stretch, QString type, QString name, int bandNumber=1);
```
Comment (jessemapel) [inline isis/src/base/objs/CubeStretch/CubeStretch.h]: This constructor doesn't appear to be implemented.

File: isis/src/base/objs/Stretch/Stretch.h [hunk 12]
```diff
@@ -71,7 +70,7 @@ namespace Isis {
    *  @history 2020-02-27 Kristin Berry - Updated to inherit from Blob so Stretches can be
    *               saved and restored from cubes. 
    */
-  class Stretch : public Isis::Blob {
```
Comment (jessemapel) [inline isis/src/base/objs/Stretch/Stretch.h:75]: This looks API breaking to me. There are functions that Stretch used to inherit from Blob that it doesn't now.

File: isis/src/base/objs/StretchBlob/StretchBlob.cpp [hunk 13]
```diff
@@ -0,0 +1,137 @@
...
+ *   software and related material nor shall the fact of distribution
+ *   constitute any such warranty, and no responsibility is assumed by the
+ *   USGS in connection therewith.
+ *
+ *   For additional information, launch
+ *   $ISISROOT/doc//documents/Disclaimers/Disclaimers.html
+ *   in a browser or see the Privacy &amp; Disclaimers page on the Isis website,
+ *   http://isis.astrogeology.usgs.gov, and the USGS privacy and disclaimers on
+ *   http://www.usgs.gov/privacy.html.
+ */
```
Comment (jessemapel) [inline isis/src/base/objs/StretchBlob/StretchBlob.cpp]: Old boilerplate

File: isis/src/base/objs/StretchBlob/StretchBlob.h [hunk 14]
```diff
@@ -0,0 +1,61 @@
...
+ *   in a browser or see the Privacy &amp; Disclaimers page on the Isis website,
+ *   http://isis.astrogeology.usgs.gov, and the USGS privacy and disclaimers on
+ *   http://www.usgs.gov/privacy.html.
+ */
+
+#include "Blob.h"
+#include "CubeStretch.h"
+
+namespace Isis {
+  class CubeStretch;
```
Comment (jessemapel) [inline isis/src/base/objs/StretchBlob/StretchBlob.h]: Why do you forward declare and include this?

File: isis/src/base/objs/StretchBlob/StretchBlob.h [hunk 15]
```diff
@@ -0,0 +1,61 @@
...
+  /**
+   * @brief Blob to store stretch information to cube.
```
Comment (jessemapel) [inline isis/src/base/objs/StretchBlob/StretchBlob.h]: This doesn't make sense. Is it "Blob to store stretch information on a cube.", "Blob to store stretch information for a cube.", or something else entirely?

File: isis/src/base/objs/StretchBlob/StretchBlob.h [hunk 16]
```diff
@@ -0,0 +1,61 @@
...
+ *
+ *   For additional information, launch
+ *   $ISISROOT/doc//documents/Disclaimers/Disclaimers.html
+ *   in a browser or see the Privacy &amp; Disclaimers page on the Isis website,
+ *   http://isis.astrogeology.usgs.gov, and the USGS privacy and disclaimers on
+ *   http://www.usgs.gov/privacy.html.
+ */
+
+#include "Blob.h"
+#include "CubeStretch.h"
```
Comment (jessemapel) [inline isis/src/base/objs/StretchBlob/StretchBlob.h:16]: QString, istream and fstream are part of the API, they should be included here or forward declared for safety sake.

File: isis/src/base/objs/StretchBlob/StretchBlob.h [hunk 17]
```diff
@@ -0,0 +1,61 @@
...
+ *   software and related material nor shall the fact of distribution
+ *   constitute any such warranty, and no responsibility is assumed by the
+ *   USGS in connection therewith.
+ *
+ *   For additional information, launch
+ *   $ISISROOT/doc//documents/Disclaimers/Disclaimers.html
+ *   in a browser or see the Privacy &amp; Disclaimers page on the Isis website,
+ *   http://isis.astrogeology.usgs.gov, and the USGS privacy and disclaimers on
+ *   http://www.usgs.gov/privacy.html.
+ */
```
Comment (jessemapel) [inline isis/src/base/objs/StretchBlob/StretchBlob.h]: old boilerplate

File: isis/src/base/objs/StretchBlob/StretchBlob.cpp [hunk 18]
```diff
@@ -0,0 +1,137 @@
...
+  /**
+   * Default constructor
+   */
+  StretchBlob::StretchBlob() : Blob("CubeStretch", "Stretch") {
+    m_stretch = new CubeStretch();
+  }
+
+
+  /**
+   * Default constructor
```
Comment (jessemapel) [inline isis/src/base/objs/StretchBlob/StretchBlob.cpp]: This is not the default constructor

File: isis/src/base/objs/StretchBlob/StretchBlob.cpp [hunk 19]
```diff
@@ -0,0 +1,137 @@
...
+   */
+  StretchBlob::StretchBlob(CubeStretch stretch) : Blob("CubeStretch", "Stretch"){
+    m_stretch = new CubeStretch(stretch);
+    Label()["Name"] = m_stretch->getName();
+    Label() += PvlKeyword("StretchType", m_stretch->getType());
+    Label() += PvlKeyword("BandNumber", QString::number(m_stretch->getBandNumber()));
```
Comment (jessemapel) [inline isis/src/base/objs/StretchBlob/StretchBlob.cpp]: Is there a reason why the Label function is called instead of the directly accessing the protected member?

File: isis/src/base/objs/StretchBlob/StretchBlob.h [hunk 20]
```diff
@@ -0,0 +1,61 @@
...
+
+      CubeStretch getStretch(); 
+
+    protected:
+      void WriteInit();
+      void ReadData(std::istream &is);
+      void WriteData(std::fstream &os);
+
+    private:
+      CubeStretch *m_stretch; //! Stretch associated with the blob
```
Comment (jessemapel) [inline isis/src/base/objs/StretchBlob/StretchBlob.h]: Can you use a smart pointer here to make this safer against memory leaks, etc?

Comment (jessemapel) [reply isis/src/base/objs/StretchBlob/StretchBlob.h]: Looking at the implementation, it doesn't look like this even needs to be a pointer.

File: isis/src/qisis/objs/CubeViewport/CubeViewport.h [hunk 21]
```diff
@@ -174,14 +175,14 @@ namespace Isis {
           const BandInfo &operator=(BandInfo other);
 
           //! @return The Stretch
-          Stretch getStretch() const;
+          CubeStretch getStretch() const;
```
Comment (jessemapel) [inline isis/src/qisis/objs/CubeViewport/CubeViewport.h:178]: This also looks API breaking

File: isis/src/qisis/objs/StretchTool/AdvancedStretch.cpp [hunk 22]
```diff
@@ -125,28 +126,24 @@ namespace Isis {
...
+    else if (stretchTypeName.compare("Manual") == 0) {
       index = 3;
     }
-    // Fail by defaulting to Linear
 
+    // Fail by defaulting to Linear. This is correct for non-Advanced Stretches anyway.
     p_stretchTypeSelection->setCurrentIndex(index);
-    StretchType *stretchType = (StretchType *)
-                               p_stretchTypeStack->currentWidget();
-    stretchType->setStretch(newStretch);
```
Comment (jessemapel) [inline isis/src/qisis/objs/StretchTool/AdvancedStretch.cpp:149]: Why was this removed, it looks like it's no longer changing the stretch in the widget.

File: isis/src/base/objs/Stretch/Stretch.h (hunk 2 above)
Comment (ladoramkershner) [reply isis/src/base/objs/Stretch/Stretch.h:94]: 👍 That all makes sense and is way more clean.

File: isis/src/qisis/objs/CubeViewport/CubeViewport.cpp [hunk 23]
```diff
@@ -1669,7 +1670,7 @@ namespace Isis {
 
 
   //! Return the gray band stretch
-  Stretch CubeViewport::grayStretch() const {
+  CubeStretch CubeViewport::grayStretch() const {
```
Comment (ladoramkershner) [inline isis/src/qisis/objs/CubeViewport/CubeViewport.cpp:1673]: There are a lot of other instances of Stretch in this file that aren't changed. For example, just after this there is CubeViewport::redStretch(), CubeViewport::greenStretch(), and CubeViewport::blueStretch(), which are all left as 'Stretch'. So I am curious why certain Typecasts are left while these ones are changed.

File: isis/src/qisis/objs/CubeViewport/CubeViewport.h [hunk 24]
```diff
@@ -339,7 +340,7 @@ namespace Isis {
        */
       double grayPixel(int sample, int line);
       //! @return The gray Stretch
-      Stretch grayStretch() const;
+      CubeStretch grayStretch() const;
       //! @return The red Stretch
       Stretch redStretch() const;
```
Comment (ladoramkershner) [inline isis/src/qisis/objs/CubeViewport/CubeViewport.h]: Again, why is grayStretch a CubeStretch and red/green/blueStretch not? Is it just that this is not intended to save stretches of RGB images?

**Review by jessemapel — COMMENTED** [review]
Sorry this took so long. This is a rather large PR so I've got a lot of questions about things. My main concern is the API. I think this is the right choice, so I guess we tick to 5.0?

File: isis/src/qisis/objs/CubeViewport/CubeViewport.cpp (hunk 23 above)
Comment (krlberry) [reply isis/src/qisis/objs/CubeViewport/CubeViewport.cpp:1673]: This was an oversight because originally this was just designed to work with gray stretches. Thanks for catching this and pointing it out!

File: isis/src/qisis/objs/CubeViewport/CubeViewport.h (hunk 24 above)
Comment (krlberry) [reply isis/src/qisis/objs/CubeViewport/CubeViewport.h]: This was an oversight because originally this was just designed to work with gray stretches. Thanks for catching this and pointing it out!

File: isis/src/base/objs/CubeStretch/CubeStretch.h (hunk 11 above)
Comment (krlberry) [reply isis/src/base/objs/CubeStretch/CubeStretch.h]: If it wasn't breaking anything, I guess it wasn't being used, so I just removed it.

File: isis/src/base/objs/StretchBlob/StretchBlob.h (hunk 15 above)
Comment (krlberry) [reply isis/src/base/objs/StretchBlob/StretchBlob.h]: I think both of your sentences make sense here and I'm not sure I'm understanding the distinction. The `StretchBlob` stores stretch information specifically associated with a Cube, but since it's a Blob also allows you to store that information on the Cube. I updated to use your second phrasing, let me know if that works for you.

File: isis/src/base/objs/StretchBlob/StretchBlob.h (hunk 14 above)
Comment (krlberry) [reply isis/src/base/objs/StretchBlob/StretchBlob.h]: Going to guess it was because in some past version of this code, it wouldn't compile without it and I didn't clean it up. Thanks for catching this! I've removed it.

File: isis/src/base/objs/StretchBlob/StretchBlob.cpp (hunk 18 above)
Comment (krlberry) [reply isis/src/base/objs/StretchBlob/StretchBlob.cpp]: Very true.

File: isis/src/base/objs/StretchBlob/StretchBlob.cpp (hunk 19 above)
Comment (krlberry) [reply isis/src/base/objs/StretchBlob/StretchBlob.cpp]: Nope! Just didn't think of it. Updated to use the protected member.

File: isis/src/qisis/objs/CubeViewport/CubeViewport.h (hunk 21 above)
Comment (krlberry) [reply isis/src/qisis/objs/CubeViewport/CubeViewport.h:178]: I agree :/ I have to admit that I wasn't thinking of API breaking things in the qisis classes, but you're right that they are part of our API.

Comment (krlberry) [reply isis/src/qisis/objs/CubeViewport/CubeViewport.h:178]: This comment is going to be resolved by releasing this code along with other API-breaking changes, noting the differences, and incrementing the ISIS version number.

File: isis/src/base/objs/Stretch/Stretch.h (hunk 12 above)
Comment (krlberry) [reply isis/src/base/objs/Stretch/Stretch.h:75]: We decided to resolve this by acknowledging that this PR is API breaking and merging it along with other API-breaking changes and incrementing the ISIS version.

File: isis/src/base/objs/CubeStretch/CubeStretch.cpp (hunk 10 above)
Comment (krlberry) [reply isis/src/base/objs/CubeStretch/CubeStretch.cpp:21]: Good point on this. Thanks! Really helped to consolidate the plethora of constructors.

File: isis/src/base/objs/StretchBlob/StretchBlob.h (hunk 15 above)
Comment (krlberry) [reply isis/src/base/objs/StretchBlob/StretchBlob.h]: I updated this again. It's both! So I updated it to say so.

File: isis/src/base/objs/Blob/Blob.cpp (hunk 5 above)
Comment (krlberry) [reply isis/src/base/objs/Blob/Blob.cpp:215]: This is handled below! It throws an exception if it's not found.

File: isis/src/qisis/objs/StretchTool/AdvancedStretch.cpp (hunk 22 above)
Comment (krlberry) [reply isis/src/qisis/objs/StretchTool/AdvancedStretch.cpp:149]: I believe this code was unnecessary in the first place. Line 146 takes care of both. On lines 83-84:

```
    connect(p_stretchTypeSelection, SIGNAL(currentIndexChanged(int)),
            p_stretchTypeStack, SLOT(setCurrentIndex(int)));
```
So, when after `setCurrentIndex(index)` is called, the index is also set on `p_stretchTypeStack` which is what changes the stretch type widget.

**Review by scsides — APPROVED** [review]

**Review by jessemapel — APPROVED** [review]
Looks good, need to wait on CI to finish still.

**Comment by krlberry** [comment]
The CI looks good except for things that are also failing on the nightly dev build, so I'm going to go ahead and merge.
